app/src/main/cpp/native-lib.cpp

```diff
@@ -12,9 +38,208 @@ Java_com_example_cloudonix_MainActivity_stringFromJNI(
+        // Iterate IPv6 addresses first
+        if ((ifaddr_iterator->ifa_addr) && (ifaddr_iterator->ifa_addr->sa_family == AF_INET6)) {
+            struct sockaddr_in6 *in6 = (struct sockaddr_in6 *) ifaddr_iterator->ifa_addr;
+            getnameinfo(ifaddr_iterator->ifa_addr, sizeof(struct sockaddr_in6), addr,
+                        sizeof(addr), NULL, 0, NI_NUMERICHOST);
+//            std::string address(ifaddr_iterator->ifa_addr->sa_data);
+            std::string address(addr);
+            if (isIPv6GlobalUnicast(address)) {
+                retVal = env->NewStringUTF(address.c_str());
+            }
+        }
+    }
+
+    // If we found nothing, keep looking...
+    if (retVal == nullptr) {
+        for (ifaddr_iterator = ifaddr;
+             ifaddr_iterator != NULL; ifaddr_iterator = ifaddr_iterator->ifa_next) {
+            // Iterate IPv4 public addresses second
+            if ((ifaddr_iterator->ifa_addr) && (ifaddr_iterator->ifa_addr->sa_family == AF_INET)) {
+                struct sockaddr_in *in = (struct sockaddr_in *) ifaddr_iterator->ifa_addr;
+                getnameinfo(ifaddr_iterator->ifa_addr, sizeof(struct sockaddr_in6), addr,
+                            sizeof(addr), NULL, 0, NI_NUMERICHOST);
+//            sockaddr *tempaddr = ifaddr_iterator->ifa_addr;
+//            std::string address(tempaddr->sa_data);
+                std::string address(addr);
+                if (isIPv4Private(address)) {
+                    retVal = env->NewStringUTF(address.c_str());
+                }
+            }
+        }
+    }
+
+    // If we found nothing, keep looking...
+    if (retVal == nullptr) {
+        //Fallback case, return any IPv4 address we found. So we can return the first one.
+        for (ifaddr_iterator = ifaddr;
+             ifaddr_iterator != NULL; ifaddr_iterator = ifaddr_iterator->ifa_next) {
+            if ((ifaddr_iterator->ifa_addr) && (ifaddr_iterator->ifa_addr->sa_family == AF_INET)) {
+                struct sockaddr_in *in = (struct sockaddr_in *) ifaddr_iterator->ifa_addr;
+                getnameinfo(ifaddr_iterator->ifa_addr, sizeof(struct sockaddr_in6), addr,
+                            sizeof(addr), NULL, 0, NI_NUMERICHOST);
+
+                std::string address(addr);
+                retVal = env->NewStringUTF(address.c_str());
+            }
+        }
+    }
+
+    // Cleanup ifaddrs
+    freeifaddrs(ifaddr);
+
+    return retVal;
+}
+
+bool isIPv6GlobalUnicast(const std::string &addr) {
+    return addr.find("2000:") == 0;
+}
+
+bool isIPv4Private(const std::string &addr) {
+    return isIPv4PrivateCase1(addr)
+           || isIPv4PrivateCase2(addr)
+           || isIPv4PrivateCase3(addr)
+           || isIPv4PrivateLinkLocal(addr);
+}
+
+bool isIPv4PrivateCase1(const std::string &addr) {
+    // If the address starts with "10." it's on the 10.x.x.x. subnet, therefore it's private
+    return addr.find("10.") == 0;
+}
+
+bool isIPv4PrivateCase2(const std::string &addr) {
```

Comment:
if we take a closer look,

`isIPv4PrivateCase2`, `isIPv4PrivateCase3` and `isIPv4PrivateLinkLocal` look strikingly similar.
copypasted even.

if these could be united in some
`checkIPv4AddressForRange(addr, int[] ranges)` and have `isIPv4Private` refactored so that it could be called as 
```
bool isIPv4Private(const std::string &addr) {
    return isIPv4PrivateCase1(addr)
           || checkIPv4AddressForRange(addr, 172, [16,31])
           || checkIPv4AddressForRange(addr, 192, [168])
           || checkIPv4AddressForRange(addr, 169, [254]); 
```
that'd be great.
but since we can't really inline arrays or lists here, the effort/cost of refactoring for a demo does not outweigth the gain.

would be great, but can always be revisited later. so i'll let the three copypasted methods slide.

Reply:
resolving.